drv/tdma_client/tdma_client.c

```diff
@@ -0,0 +1,522 @@
+            }
+        }
+    }
+}
+
+/**
+ * @brief Interruption handler for the TX state machine timer
+ *
+ */
+static void timer_tx_interrupt(void) {
+
+    bool packet_sent = false;
+
+    // Check the state of the device.
+    if (_tdma_client_vars.registration_flag == DB_TDMA_CLIENT_REGISTERED) {
+
+        // Prepare right now the next timer interruption
+        db_timer_hf_set_oneshot_us(TDMA_CLIENT_TIMER_HF, TDMA_CLIENT_HF_TIMER_CC_TX, _tdma_client_vars.tdma_client_table.frame_duration, &timer_tx_interrupt);
+
+        // send messages if available
+        packet_sent = _message_rb_tx_queue(_tdma_client_vars.tdma_client_table.tx_duration);
+
+        // if no packet has been sent for a while, send a keep_alive ping to maintain the connection.
+        if (!packet_sent) {
+            if (db_timer_hf_now(TDMA_CLIENT_TIMER_HF) - _tdma_client_vars.last_tx_packet_timestamp > TDMA_CLIENT_MAX_DELAY_WITHOUT_TX) {
+
+                _tx_keep_alive_message();
+                // Save the timestamp of the last packet
+                _tdma_client_vars.last_tx_packet_timestamp = db_timer_hf_now(TDMA_CLIENT_TIMER_HF);
+            }
+        }
+    } else {  // Device is unregistered
+
+        // Prepare right now the next timer interruption.
+        uint32_t delay_time = _get_random_delay_us();
+        db_timer_hf_set_oneshot_us(TDMA_CLIENT_TIMER_HF, TDMA_CLIENT_HF_TIMER_CC_TX, delay_time, &timer_tx_interrupt);
+
+        // Try to register with the TDMA server
+        _tx_tdma_register_message();
+        // Save the timestamp of the last packet
+        _tdma_client_vars.last_tx_packet_timestamp = db_timer_hf_now(TDMA_CLIENT_TIMER_HF);
+    }
+}
+
+/**
+ * @brief Interruption handler for the RX state machine timer
+ *
+ */
+static void timer_rx_interrupt(void) {
+
+    // If the duration of the RX timer is equal to the frame duration
+    // just leave the radio ON permanently
+
+    if (_tdma_client_vars.tdma_client_table.rx_duration == _tdma_client_vars.tdma_client_table.frame_duration) {
+        db_radio_rx();
+
+        return;
+
+    } else {
+        // Check if we just came from an idle period or an rx period
+        if (_tdma_client_vars.rx_flag == DB_TDMA_CLIENT_RX_WAIT) {
+            // Set the next interruption
+            db_timer_hf_set_oneshot_us(TDMA_CLIENT_TIMER_HF, TDMA_CLIENT_HF_TIMER_CC_RX, _tdma_client_vars.tdma_client_table.rx_duration, &timer_rx_interrupt);
+            // turn the radio ON
+            db_radio_rx();
+        } else if (_tdma_client_vars.rx_flag == DB_TDMA_CLIENT_RX_ON) {
+            // Set the next interruption
+            uint32_t delay = _tdma_client_vars.tdma_client_table.frame_duration - _tdma_client_vars.tdma_client_table.rx_duration;
+            db_timer_hf_set_oneshot_us(TDMA_CLIENT_TIMER_HF, TDMA_CLIENT_HF_TIMER_CC_RX, delay, &timer_rx_interrupt);
+            // turn the radio OFF
+            db_radio_disable();
+        }
+    }
```

Comment:
No need for an else clause since you return in the if block
```suggestion
    }

    // Check if we just came from an idle period or an rx period
    if (_tdma_client_vars.rx_flag == DB_TDMA_CLIENT_RX_WAIT) {
        // Set the next interruption
        db_timer_hf_set_oneshot_us(TDMA_CLIENT_TIMER_HF, TDMA_CLIENT_HF_TIMER_CC_RX, _tdma_client_vars.tdma_client_table.rx_duration, &timer_rx_interrupt);
        // turn the radio ON
        db_radio_rx();
    } else if (_tdma_client_vars.rx_flag == DB_TDMA_CLIENT_RX_ON) {
        // Set the next interruption
        uint32_t delay = _tdma_client_vars.tdma_client_table.frame_duration - _tdma_client_vars.tdma_client_table.rx_duration;
        db_timer_hf_set_oneshot_us(TDMA_CLIENT_TIMER_HF, TDMA_CLIENT_HF_TIMER_CC_RX, delay, &timer_rx_interrupt);
        // turn the radio OFF
       db_radio_disable();
    }
```

Reply:
done!